File: src/publisher/Publisher.go
package publisher

import (
	"fmt"
	zmq "github.com/pebbe/zmq4"
)

type PublisherI interface {
	Connect(endpoint string) error
	Put(topic string, msg string)
	Destroy() error
}

type Publisher struct {
	zctx *zmq.Context
	s    *zmq.Socket
}

func NewPublisher(zctx *zmq.Context) (*Publisher, error) {
	p := new(Publisher)
	p.zctx = zctx
	s, e := p.zctx.NewSocket(zmq.REQ)
	p.s = s
	return p, e
}

func (p *Publisher) Destroy() error {
	return p.s.Close()
}

func (p *Publisher) Connect(endpoint string) error {
	return p.s.Connect(endpoint)
}

func (p *Publisher) Put(topic string, msg string) {
	p.s.SendMessage(topic + " " + msg)
	fmt.Printf("Put mandou: %s %s\n", topic, msg)
	reply, _ := p.s.RecvMessage(0)
	fmt.Printf("Put recebeu: %s\n", reply)
}

File: src/main/java/publisher/Publisher.java
package publisher;

import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZMsg;

public class Publisher {
    private Socket socket;

    public Publisher(ZContext zctx) {
        this.socket = zctx.createSocket(SocketType.REQ);
    }

    public void destroy() {
        this.socket.close();
    }

    public boolean connect(String endpoint) {
        return this.socket.connect(endpoint);
    }

    public boolean put(String topic, String msg) {
        ZMsg zmsg = new ZMsg();
        zmsg.add(topic);
        zmsg.add(msg);

        if (!zmsg.send(this.socket))
            return false;
        zmsg = ZMsg.recvMsg(this.socket);

        System.out.printf("Put success: %s", zmsg.getLast());
        return true;
    }
}
